pkg/services/ngalert/schedule/schedule.go

```diff
@@ -230,9 +230,20 @@ func (sch *schedule) processTick(ctx context.Context, dispatcherGroup *errgroup.
 	sch.metrics.SchedulableAlertRulesHash.Set(float64(hashUIDs(alertRules)))
 
 	readyToRun := make([]readyToRunItem, 0)
+	pausedAlertRuleKeys := make(map[ngmodels.AlertRuleKey]struct{}, 0)
 	missingFolder := make(map[string][]string)
 	for _, item := range alertRules {
 		key := item.GetKey()
+		if item.IsPaused {
+			if ruleInfo, ok := sch.registry.del(key); ok {
```

Comment:
I just realized one subtle problem that you will not catch unless the change is made via API. When you update a rule via API, the handler calls Scheduler.UpdateAlertRule that basically sends a message to the `updateCh` of the rule evaluation routine, which resets all states. Therefore, by the time the scheduler executes this path, the state will be gone!

Reply:
What if we just skip the state cleaning from the routine and just wait for the next tick to clean the state and the routine registry?

Reply:
Like this:
```go
select {
    // used by external services (API) to notify that rule is updated.
    case ctx := <-updateCh:
        // we can continue directly if the rule is paused because in the next tick it will be picked up. The state
        // will be cleaned up and the routine will be stopped and removed from the registry
        if ctx.IsPaused {
            logger.Info("Skip updating rule because it is paused")
            continue
        }
        ...
```

I have tested it and it works perfectly. The only thing I don't like is the 2-step pausing flow but it quite simple.

Reply:
@yuri-tceretian I'm not sure I am following. Could you help me understand the issue?

1. When the API calls `UpdateAlertRule` it sends a `ruleVersion` into `updateCh`.
2. This then calls `clearState(grafanCtx, errRuleDeleted)`. However, at this time, the rule still exists in the `alertRuleInfoRegistry`, 
3. The next tick occurs, and the scheduler gets the latest versions of alert rules from the database.
4. The scheduler sees that the rule has been paused, deletes it from the registry, and then calls `ruleInfo.Stop(errRulePaused)`
5. The context should be canceled with `errRulePaused` which then calls `clearState(ctx, grafanaCtx.Err())` a second time, but this time with `errRulePaused` instead of `errRuleDeleted`

Reply:
The problem is that the first time we clear the state we don't save the state's history (because `clearState` is used with `errRuleDeleted`). That means that when we go to clear the state the 2nd time there won't be any state to save inside the history.

Reply:
so we have 2 options:
- skip the clearState within the updateCh flow
or
- clear the state within the updateCh with the `errRulePaused`

I prefer the first option as we save a call to `clearState` (saving calls to DB)